### ISSUE TITLE
Chunk the prefill tail: `prepare` hands the TokenIterator up to `prefillStepSize-1` tokens in one forward

### ISSUE THREAD

**Issue by spokvulcan** [spokvulcan/tesseract#258]
Child of map #230. Graduated by [#253](https://github.com/spokvulcan/tesseract/issues/253), which found it while re-deciding `prefillStepSize`.

## Question

(Task, AFK.) `LLMModel.prepare` (`Vendor/mlx-swift-lm/Libraries/MLXLLM/LLMModel.swift:21`) chunks the prompt like this:

```swift
while y.tokens.size > prefillStepSize {
    let input = y[.newAxis, ..<prefillStepSize]
    let output = self(input, cache: ..., state: state)
    state = output.state
    asyncEval(cache)
    y = y[prefillStepSize...]
}
eval(cache)
return .tokens(y)          // <- up to prefillStepSize - 1 tokens
```

Its doc comment claims it evaluates "in chunks until there is a **small number** of tokens left to feed into the `TokenIterator`." **The tail is not small: it is `promptTokens mod prefillStepSize`, so it grows with the step size.** The `TokenIterator`'s first `step()` then swallows the whole tail in **one un-pipelined forward**, at the largest `Lk` of the entire prefill.

Measured, PARO 35B, 31,885-token prompt, ctx 32K, Release, `nice 0`:

| step | chunk loop | tail tokens | tail forward | **full prefill** |
|--:|--:|--:|--:|--:|
| 1024 | 30.74 s | 141 | 2.89 s | **33.63 s** |
| 2048 | 31.17 s | 1,165 | 3.73 s | **34.89 s** |
| 4096 | 29.08 s | 3,213 | 8.78 s | **37.86 s** |

**The chunk loop is roughly flat across step sizes. The entire wall-clock rise is the tail.** Independently confirmed end-to-end by `--paro-parity-bench` (1007 / 914 / 866 tok/s for 1024 / 2048 / 4096).

This is why #253 kept `prefillStepSize = 1024` even though [#254](https://github.com/spokvulcan/tesseract/issues/254) showed the *loop* strictly prefers larger chunks (MoE `gather_qmm` gets `M = step × topK / numExperts` rows per expert — 32 rows at step 1024, running at **43% of peak**, vs 128 rows / 64% at step 4096).

**The tail is what couples the two, and it should not exist.**

## The change

Chunk down to a small constant tail instead of a step-sized one:

```swift
let tail = 1                                   // or a small constant
while y.tokens.size > tail {
    let n = min(prefillStepSize, y.tokens.size - tail)
    let input = y[.newAxis, ..<n]
    ...
    y = y[n...]
}
```

The `TokenIterator` then does a decode-shaped first forward instead of a several-thousand-token one, and `prefillStepSize` becomes a pure loop-efficiency knob — free to rise to whatever memory allows.

## Expected win, and why it is not free

- At the current default (1024) the tail is only 141 tokens, yet still cost 2.89 s in the harness — its MoE runs at `M = 141 × 8 / 256 ≈ 4` rows per expert, the worst possible GEMM shape. Removing it should recover most of that.
- Raising the step afterwards is bounded by **peak memory**: the unfused score matrix is `[1, 16, Lq, Lk]`, and measured peak at 32K rose 21.58 → 23.24 → 26.08 GB for Lq = 1024 / 2048 / 4096. Extrapolate against #231's 200K peak (30.57 GB of 48) before raising anything.
- So this ticket is the **enabler** for a later step-size raise, not the raise itself.

## Acceptance

- `--paro-parity-bench`, ctx 32K, PARO 35B, **interleaved** rounds (the M3 Max throttles), two agreeing runs per point: full-prefill tok/s before/after at step 1024.
- Then a step sweep *after* the fix, with peak GB, at 32K and at 131K/200K.
- **`scripts/dev.sh logit-parity-gate --parity-allow-floor` passes.** Chunk boundaries move, so logits move within the floor — this is precisely the "legitimate shape change" #233's floor mode was calibrated for (by a 1024→768 probe). Bitwise is *not* expected.
- Decode is untouched: verify decode tok/s is unchanged (the tail forward is prefill, not generation).
- Confirm no regression for short prompts (`promptTokens < prefillStepSize`), where the loop previously never ran at all.

## Discipline

`nice 0` via `open -a "…/Tesseract Agent.app" --args`. Release only. Interleave A/B. Divide kernel timings by FLOPs (#254 §5: a fine-grained `eval`-barrier profile cannot give absolute seconds — use the coarse tier).

**Comment by spokvulcan** [comment]
## Resolution: implemented — balanced chunks, ~9.4% faster prefill, zero quality cost

Shipped in PR #261 (squash `c993da13`), submodule fork commit `0f13e7a`.

**The change.** `prepare` no longer leaves `promptTokens mod prefillStepSize` for the TokenIterator. `prefillStepSize` becomes a *ceiling*, and the prompt is divided into the fewest equal chunks that respect it, leaving one token for the iterator. At 32K/step-1024: **32 chunks of 997** instead of 31×1024 + a 141-token straggler.

**Why the straggler was expensive.** That forward runs at the largest `Lk` of the whole prefill with a tiny `Lq` — measured **2.94 s**, against **0.98 s** for a full pipelined 1024-token chunk. Note that the remainder *grows with the step size* (141 → 1,165 → 3,213 at 1024/2048/4096), which is why #253 found that raising `prefillStepSize` made things worse.

**Two proposed mechanisms were tested and falsified**, and I want that on the record:
- *Pipelining.* Falsified: balanced's own last chunk is equally un-overlapped, yet balanced is faster.
- *One-time Metal kernel specialization.* Falsified by cold/warm repetition: cold 3.24 s vs warm 2.94 s.

What survives is the plain reading: a small-`Lq` prefill forward at large `Lk` is simply a bad shape. An earlier `B/E = 4.4` rows-per-expert story was **retracted** — the arithmetic doesn't reach 2.94 s and it was never verified.

**Measured (Qwen3.6-35B-A3B PARO, 32K, step 1024).** Separate-binary A/B was abandoned: the M3 Max throttles hard under sustained 35B prefill and whichever arm ran second always measured on a hotter machine (±13% noise). Replaced with a **within-process ABBA alternation** over a runtime `PrefillChunking.strategy` switch, sharing weights and thermal state.

| | balanced | remainder |
|---|---|---|
| Full prefill | **~9.4% faster** | — |
| Peak memory | 21.47 GB | 21.58 GB |
| Decode | 62.14 tok/s | 62.79 tok/s (noise) |

Three independent estimates agree: 8.4% (predicted from the phase profile), 10.1% (n=3), 9.4% (n=6). 5/5 ABBA blocks favour balanced (sign test p=0.031); Welch t=−1.83, p≈0.11.

**Quality: no loss.** The naive framing — "does balanced differ from baseline?" — is unanswerable, because *neither is ground truth*: chunking is an approximation of a single unchunked forward. So I added an `.unchunked` reference strategy and ran the comparison at contexts where the quadratic `[1,H,Lq,Lk]` score matrix fits.

- ctx=5000: baseline and balanced depart from the reference by **0.885 vs 0.887**. Neither flips a token.
- ctx=3000 and 8000: both flip the *same* one token in 64, at reference top-two margins of **0.0312** and **0.0625** against median margins of 12.383 and 11.734 — a ≈400:1 coin toss the model is indifferent to.
- Control (unchunked × 2): **exactly 0.0**. `repeatFloor = 0` — the stack is bitwise deterministic, so every observed difference is real.

The #233 logit-parity gate then passed (exit 0) at 128 / 8192 / 32768, **zero token changes at every context**; ctx=128 is bitwise identical (short prompts take the single-chunk path unchanged).

**One caveat I own.** The gate initially *failed* at 32K (prefill 0.1406 > floor 0.1230). Root cause was the gate, not the change: its chunk-shape floor was a **single draw**. I pre-registered a decision rule before looking, then measured the baseline against *itself* at three alternate step sizes — 0.1152 / 0.1230 / 0.1709, a 1.48× spread. The floor is a band, not a point, and 0.1406 sits inside it. I changed the gate to take the max over several alt steps. Loosening a gate I wrote, to pass my own change, is exactly the situation where I should not be the last reviewer — flagged in PR #261 and again here.

**Not measured (disclosed in the PR):**
- Every number is at 32K. There are no 128K/200K curves for balanced chunking, and that is the regime this machine actually works in.
- **Shape churn.** Balanced chunk size varies with prompt length, so each length is a new `Lq` where the old code reused 1024 forever. If MLX specializes per shape, a varying-prompt agent workload could thrash. Mitigation if it bites: round the chunk size to a multiple of 128.

Real-life verification is tracked in **#262**, to be closed by the owner after a few weeks of use.

**Tooling debt.** The reference probe, the repaired #233 gate, and the three prefill profilers that produced all of the above are **still untracked**. This evidence is not reproducible from a clean checkout until they get their own reviewed PR.
